### PR TITLE
Add cloudflared tunnel ready command

### LINKED ISSUE #1126: Built-in Docker Healthcheck
**Describe the feature you'd like**
I would like the docker image to contain curl or wget so it's possible to create a docker healthcheck that docker-compose or docker swarm can use.

It would be nice the healthchecking is just native to the cloudflare/cloudflared image like so.
```Dockerfile
FROM cloudflare/cloudflared:2023.10.0 as cloudflared
FROM alpine:3.18.4

COPY --from=cloudflared /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
COPY --from=cloudflared /usr/local/bin/cloudflared /usr/local/bin

ENV TUNNEL_METRICS="localhost:2000"

HEALTHCHECK --interval=5s --retries=6 --timeout=3s CMD wget -q ${TUNNEL_METRICS}/ready -O -
```

If building from alpine is something you don't want to do, then provide a `cloudflare tunnel health` command that can be used like so
```Dockerfile
HEALTHCHECK --interval=5s --retries=6 --timeout=3s CMD cloudflare tunnel health --ready-endpoint localhost:2000/ready
```
or like so
```yml
services:
  cloudflared:
    image: cloudflare/cloudflared
    environment:
      TUNNEL_METRICS: "localhost:2000"
    healthcheck:
      command: cloudflared tunnel health --ready-endpoint ${TUNNEL_METRICS}/ready
      interval: 5s
      retries: 6
      timeout: 3s
```

**Describe alternatives you've considered**
Build my own image built on-top of the cloudflare/cloudflared image or use third-party images built on top of cloudflare's

### PR DESCRIPTION
**Pull request by firecow**
This would make it possible to have a Docker image without wget and curl, and still make docker healthchecking in docker possible

closes #1126

### CODE REVIEW

File: cmd/cloudflared/tunnel/subcommands.go
Comment (jlaffaye) [inline cmd/cloudflared/tunnel/subcommands.go:415]: Why not 
```suggestion
	requestURL := fmt.Sprintf("http://%s/ready", c.String("metrics")
```
?

Comment (firecow) [reply cmd/cloudflared/tunnel/subcommands.go:415]: Yeah, we not :smile:

**Comment by firecow** [comment]
@jlaffaye I there anything I can do to fast-track the PR somehow ?

File: cmd/cloudflared/tunnel/subcommands.go
Comment (felipelima-circle) [inline cmd/cloudflared/tunnel/subcommands.go:414]: I think the `/healthcheck` endpoint would be more appropriate.

Comment (firecow) [reply cmd/cloudflared/tunnel/subcommands.go:414]: I wasn't aware that such an endpoint existed :+1: I will change it. @felipelima-circle

**Comment by firecow** [comment]
@felipelima-circle 

image

Am I to understand that the /healthcheck "just" returns the string "ok"? 
It make no checking if the tunnels are actually healthy?

**Comment by felipelima-circle** [comment]
@firecow That makes sense. 👍🏻

**Comment by CodaBool** [comment]
A merge of master into your branch may be in order because I was getting an error about "UDP Buffer Sizes" that forced me to merge in master. Which resolved it. But past that this seems to work fine on a quick test.


I stood up a separate service at port 8081 to make a cloudflare tunnel with. Then wrote a docker healthcheck into a docker-compose.yml. Which had my local IP there:
```yaml
healthcheck:
  test: tunnel --metrics 192.168.0.15:8081 ready || exit 1
```

Which worked great. The was a little high at 30 seconds but it's fine.

**Comment by firecow** [comment]
> A merge of master into your branch may be in order because I was getting an error about "UDP Buffer Sizes" that forced me to merge in master. Which resolved it. But past that this seems to work fine on a quick test.
> 
> I stood up a separate service at port 8081 to make a cloudflare tunnel with. Then wrote a docker healthcheck into a docker-compose.yml. Which had my local IP there:
> 
> ```yaml
> healthcheck:
>   test: tunnel --metrics 192.168.0.15:8081 ready || exit 1
> ```
> 
> Which worked great. The timeout was a little high at 30 seconds but it's fine.

Master has been merged.

**Comment by kamaradski** [comment]
I'm really waiting for this, as in AWS ECS I can also not get the workaround (mentioned here) working at all. If something breaks our tunnels are currently just hanging. not ideal.

**Comment by firecow** [comment]
@DevinCarr or @jcsf One of you guys, wanna give this PR some :heart: ?

**Comment by firecow** [comment]
@chungthuang, @DevinCarr and @jcsf  I've synced the fork

**Comment by huyz** [comment]
Can anyone push this further along?

**Comment by huyz** [comment]
What would be the use case of the old /heathcheck endpoint, if any?

**Comment by huyz** [comment]
Is there a docker image for this, while we wait for an official merge?

**Comment by firecow** [comment]
@huyz I've just synced the fork. But we need some maintainer or developer to approve the test runs, and eventually merge the change into master.

File: cmd/cloudflared/tunnel/subcommands.go
Comment (chungthuang) [inline cmd/cloudflared/tunnel/subcommands.go:415]: The default value for `--metrics` is `localhost:` which doesn't have the port. Can you add a check here to make sure the caller actually provides the metrics address and port?
```
if !c.IsSet("metrics") {
    return fmt.Errorf("--metrics has to be provided")
}
```

Comment (firecow) [reply cmd/cloudflared/tunnel/subcommands.go:415]: Done :+1:

**Comment by firecow** [comment]
```
~/workspace/firecow-cloudflared-fork$ make cloudflared && ./cloudflared tunnel --metrics 127.0.0.1:40026 ready; echo $?
GOOS=linux GOARCH=amd64  go build -mod=vendor  -ldflags='-X "main.Version=c39f0ae3" -X "main.BuildTime=2024-11-14-1739 UTC" ' github.com/cloudflare/cloudflared/cmd/cloudflared
Get "http://127.0.0.1:40026/ready": dial tcp 127.0.0.1:40026: connect: connection refused
1
~/workspace/firecow-cloudflared-fork$ make cloudflared && ./cloudflared tunnel --metrics 127.0.0.1:40027 ready; echo $?
GOOS=linux GOARCH=amd64  go build -mod=vendor  -ldflags='-X "main.Version=c39f0ae3" -X "main.BuildTime=2024-11-14-1739 UTC" ' github.com/cloudflare/cloudflared/cmd/cloudflared
0
```

Just leaving this here for reference.

It seems like the feature is doing what it is supposed to do.

**Review by chungthuang — APPROVED** [review]

**Comment by huyz** [comment]
For a noob, can someone give the exact syntax to put in `docker-compose.yml` to make this healthcheck work out of the box?

**Comment by IngmarStein** [comment]
@huyz here's what I use:

```
    environment:
      - TUNNEL_TOKEN=${TUNNEL_TOKEN}
      - TUNNEL_METRICS=0.0.0.0:60123
    healthcheck:
      test: ["CMD", "cloudflared", "tunnel", "--metrics", "localhost:60123", "ready"]
      interval: 30s
      timeout: 30s
      retries: 3
```

**Pull request merged**
